Additional Steps for BQ_Source

Praveena2607 opened the pull request:
Please review my Scenarios.

masivesky reviewed (commented):
Please update the PR description to reflect details and context about the changes. For eg: https://github.com/data-integrations/google-cloud/pull/1444#issue-2518603891

masivesky reviewed (commented):
The BUTs seems to be failing, please take a look and fix the errors.

Praveena2607 commented:
**Description:**

Updated the BigQuery plugin to include additional steps for better handling.
Modified property locators in Cdfplugin property locators.java to support new BigQuery functionality.
Updated the errorMessage.properties and pluginParameters.properties file to ensure proper error handling and parameter management for the new configuration.

**Modified Files:**

Cdfplugin property locators.java
errorMessage.properties
pluginParameters.properties

**File Modified:**

src/e2e-test/java/io/cdap/plugin/utils/CdfPluginPropertyLocator.java

**Unit Tests:**

No changes made to unit tests.
Verified that all new test cases pass successfully.

masivesky commented:
Build with unit tests is still failing, there are a bunch of checkstyle errors, for eg:

```
/tmp/google-cloud/google-cloud/src/e2e-test/java/io/cdap/plugin/utils/E2ETestConstants.java:20:68: '=' is not preceded with whitespace. [WhitespaceAround]
```

Please check and fix the same

itsankit-google commented on pom.xml:
```diff
@@ -86,7 +86,7 @@
     <google.cloud.datastore.version>1.105.1</google.cloud.datastore.version>
     <google.protobuf.java.version>3.19.4</google.protobuf.java.version>
-    <google.tink.version>1.3.0-rc3</google.tink.version>
+    <google.tink.version>1.5.0</google.tink.version>
```
why this change?

Praveena2607 replied:
I encountered dependency errors while running mvn clean install after resolving all the checkstyle issues. The build was failing due to the outdated version of google.tink.version. Updating it to 1.5.0 resolved the issue, and no other changes were made.

itsankit-google replied:
please do not make dependency changes in the e2e tests PR.

Praveena2607 replied:
Refactor CdfPluginPropertyLocator alignment and update build dependencies.
-Fixed indentation alignment for partitionFrom, partitionTo, filter and Output Schema.
-Updated pom.xml as it is.

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/utils/CdfPluginPropertyLocator.java:
```diff
@@ -46,35 +51,39 @@ public enum CdfPluginPropertyLocator {
-      .put("gcsMoveSourcePath", CdfPluginPropertyLocator.GCS_MOVE_SOURCE_PATH)
-      .put("gcsMoveDestinationPath", CdfPluginPropertyLocator.GCS_MOVE_DESTINATION_PATH)
-      .build();
+            .put("projectId", CdfPluginPropertyLocator.PROJECT_ID)
+            .put("datasetProjectId", CdfPluginPropertyLocator.DATASET_PROJECT_ID)
+            .put("dataset", CdfPluginPropertyLocator.DATASET)
+            .put("table", CdfPluginPropertyLocator.TABLE)
+            .put("format", CdfPluginPropertyLocator.FORMAT)
+            .put("path", CdfPluginPropertyLocator.PATH)
+            .put("sampleSize", CdfPluginPropertyLocator.SAMPLE_SIZE)
+            .put("delimiter", CdfPluginPropertyLocator.DELIMITER)
+            .put("skipHeader", CdfPluginPropertyLocator.SKIP_HEADER)
+            .put("pathSuffix", CdfPluginPropertyLocator.SUFFIX)
+            .put("encryptionKeyName", CdfPluginPropertyLocator.CMEK_KEY)
+            .put("serviceAccountType", CdfPluginPropertyLocator.SERVICE_ACCOUNT_TYPE)
+            .put("serviceAccountFilePath", CdfPluginPropertyLocator.SERVICE_ACCOUNT_PATH)
+            .put("serviceAccountJSON", CdfPluginPropertyLocator.SERVICE_ACCOUNT_JSON)
+            .put("truncateTable", CdfPluginPropertyLocator.TRUNCATE_TABLE)
+            .put("updateTableSchema", CdfPluginPropertyLocator.UPDATE_TABLE_SCHEMA)
+            .put("topic", CdfPluginPropertyLocator.PUBSUB_TOPIC)
+            .put("maximumBatchCount", CdfPluginPropertyLocator.PUBSUB_MAXIMUM_BATCH_COUNT)
+            .put("maximumBatchSize", CdfPluginPropertyLocator.PUBSUB_MAXIMUM_BATCH_SIZE)
+            .put("publishDelayThreshold", CdfPluginPropertyLocator.PUBSUB_PUBLISH_DELAY_THRESHOLD)
+            .put("retryTimeout", CdfPluginPropertyLocator.PUBSUB_RETRY_TIMEOUT)
+            .put("errorThreshold", CdfPluginPropertyLocator.PUBSUB_ERROR_THRESHOLD)
+            .put("outputSchema", CdfPluginPropertyLocator.OUTPUT_SCHEMA_MACRO_INPUT)
+            .put("objectsToDelete", CdfPluginPropertyLocator.GCS_DELETE_OBJECTS_TO_DELETE)
+            .put("objectsToCreate", CdfPluginPropertyLocator.GCS_CREATE_OBJECTS_TO_CREATE)
+            .put("createFailIfObjectExists", CdfPluginPropertyLocator.GCS_CREATE_FAIL_IF_OBJECT_EXISTS)
+            .put("gcsMoveSourcePath", CdfPluginPropertyLocator.GCS_MOVE_SOURCE_PATH)
+            .put("gcsMoveDestinationPath", CdfPluginPropertyLocator.GCS_MOVE_DESTINATION_PATH)
+            .put("partitionFrom", CdfPluginPropertyLocator.PARTITION_START_DATE)
+            .put("partitionTo", CdfPluginPropertyLocator.PARTITION_END_DATE)
+            .put("filter", CdfPluginPropertyLocator.FILTER)
+            .put("Output Schema-macro-input", CdfPluginPropertyLocator.OUTPUT_SCHEMA)
+            .build();
```
why this change?

Praveena2607 replied:
I was unable to add my steps properly in this field because the alignments were incorrect. To fix this, I copied and pasted the same content and added my steps without making any other changes.

itsankit-google replied:
please fix the code style in your editor.

Praveena2607 replied:
Refactor CdfPluginPropertyLocator alignment and update build dependencies.
-Fixed indentation alignment for partitionFrom, partitionTo, filter and Output Schema.
-Updated pom.xml as it is.

itsmekumari commented on src/e2e-test/features/bigquery/source/BigQuerySourceError.feature:
```diff
@@ -55,3 +55,24 @@ Feature: BigQuery source - Validate BigQuery source plugin error scenarios
+    Then Replace input plugin property: "table" with value: "bqSourceTable"
+    Then Click on the Get Schema button
+    Then Enter BigQuery source properties partitionFrom and partitionTo
+    Then Validate BigQuery source incorrect property error for Partition Start date "<property>" value "<value>"
```
@Praveena2607 I don't think the internal review comments have been addressed. Please address all the internal review comments provided earlier in this PR as well. Ref https://github.com/Praveena2607/google-cloud/pull/1

itsmekumari replied:
As discussed over call, Please resolve all the discussed points and mentioned in internal PR. To use the latest framework steps in all scenarios, for entering the values in property, for validating inline error messages, for selecting output schema macro.

itsmekumari commented on src/e2e-test/features/bigquery/source/BigQueryToGCS_WithMacro.feature:
```diff
@@ -69,3 +69,222 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
+    Then Validate the cmek key "cmekGCS" of target GCS bucket if cmek is enabled
+
+  @CMEK @BQ_SOURCE_TEST @GCS_SINK_TEST
+  Scenario:Validate successful records transfer from BigQuery to GCS with macro arguments for filter and outputschema
```
Output schema macro step is not present in scenario but in title it is written. Make sure if not using in scenario remove from title, and add in the scenario where step is used.

itsmekumari commented on src/e2e-test/features/bigquery/source/BigQueryToGCS_WithMacro.feature:
```diff
@@ -69,3 +69,222 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
+  Scenario:Validate successful records transfer from BigQuery to GCS with macro arguments for filter and outputschema
+    Given Open Datafusion Project to configure pipeline
+    When Source is BigQuery
+    When Sink is GCS
```
Use the latest steps from framework.

itsmekumari commented on src/e2e-test/features/bigquery/source/BigQueryToGCS_WithMacro.feature:
```diff
@@ -69,3 +69,222 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
+    Then Enter BigQuery property "serviceAccountJSON" as macro argument "serviceAccount"
+    Then Enter BigQuery property "dataset" as macro argument "bqDataset"
+    Then Enter BigQuery property "table" as macro argument "bqSourceTable"
+    Then Enter BigQuery source property output schema "outputSchema" as macro argument "bqOutputSchema"
```
Use the existing step discussed earlier, for selecting macro action for output schema.

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:
```diff
@@ -260,4 +264,124 @@ public void validateRecordsTransferredToTargetTableIsEqualToNumberOfRecordsFromS
     Assert.assertEquals(count, countRecordsTarget);
   }
+
+  @Then("Enter BigQuery source properties partitionFrom and partitionTo")
```
No need to add the step definition. Please make use of existing steps. This was discussed may times in our calls.

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:
```diff
@@ -260,4 +264,124 @@ public void validateRecordsTransferredToTargetTableIsEqualToNumberOfRecordsFromS
+    CdfBigQueryPropertiesActions.enterPartitionEndDate(new SimpleDateFormat("dd-MM-yyyy")
+            .format(DateUtils.addDays(new Date(), 1)));
+  }
+  @Then("Validate BigQuery source incorrect property error for Partition Start date {string} value {string}")
```
No need to add the step definition. Please make use of existing steps.

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:
```diff
@@ -260,4 +264,124 @@ public void validateRecordsTransferredToTargetTableIsEqualToNumberOfRecordsFromS
+    Assert.assertEquals(expectedColor, actualColor);
+  }
+
+  @Then("Validate BigQuery source incorrect property error for Partition End date {string} value {string}")
```
Here as well.

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:
```diff
@@ -260,4 +264,124 @@ public void validateRecordsTransferredToTargetTableIsEqualToNumberOfRecordsFromS
+    Assert.assertEquals(expectedColor, actualColor);
+  }
+
+  @Then("Validate BigQuery source incorrect property error for reference name{string} value {string}")
```
Here as well.

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:
```diff
@@ -260,4 +264,124 @@ public void validateRecordsTransferredToTargetTableIsEqualToNumberOfRecordsFromS
+
+  }
+
+  @Then("Enter BigQuery source properties filter")
```
All the step definitions added here in base file are all not needed. Use the existing steps, and execute the tests.

itsankit-google commented on pom.xml:
```diff
@@ -86,7 +86,7 @@
     <google.cloud.datastore.version>1.105.1</google.cloud.datastore.version>
     <google.protobuf.java.version>3.19.4</google.protobuf.java.version>
-    <google.tink.version>1.3.0-rc3</google.tink.version>
+    <google.tink.version>1.3.0</google.tink.version>
```
this is still not reverted to older version

Praveena2607 replied:
Now, It is reverted and please review.

itsankit-google reviewed (changes requested):
please revert back the checkstyle changes in existing steps.

masivesky commented on src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:
```diff
@@ -16,25 +16,31 @@
 package io.cdap.plugin.bigquery.stepsdesign;
 
 import io.cdap.e2e.pages.actions.CdfBigQueryPropertiesActions;
+import io.cdap.e2e.pages.actions.CdfPluginPropertiesActions;
 import io.cdap.e2e.pages.actions.CdfStudioActions;
+import io.cdap.e2e.pages.locators.CdfBigQueryPropertiesLocators;
 import io.cdap.e2e.pages.locators.CdfStudioLocators;
 import io.cdap.e2e.utils.BigQueryClient;
 import io.cdap.e2e.utils.ConstantsUtil;
 import io.cdap.e2e.utils.ElementHelper;
 import io.cdap.e2e.utils.PluginPropertyUtils;
 import io.cdap.e2e.utils.SeleniumHelper;
+import io.cdap.e2e.utils.WaitHelper;
 import io.cdap.plugin.common.stepsdesign.TestSetupHooks;
+import io.cdap.plugin.utils.CdfPluginPropertyLocator;
 import io.cdap.plugin.utils.E2EHelper;
 import io.cdap.plugin.utils.E2ETestConstants;
 import io.cucumber.java.en.Then;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.time.DateUtils;
 import org.junit.Assert;
 import stepsdesign.BeforeActions;
 
 import java.io.IOException;
 import java.text.SimpleDateFormat;
 import java.util.Date;
 import java.util.Optional;
+import java.util.Properties;
```
Any reason we are adding additional imports here? Don't see them being used in the file anywhere

Praveena2607 replied:
Thank you for pointing this out. I did not manually add any of these imports. This branch was created by cloning from the master, and it seems these additional imports were automatically included during the branching process. could you please suggest how we should proceed?

Praveena2607 replied:
I have now removed the unnecessary imports that are not in use.

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/utils/E2ETestConstants.java:
```diff
@@ -17,4 +17,8 @@ public class E2ETestConstants {
   public static final String ERROR_MSG_BQ_INCORRECT_CHUNKSIZE = "errorMessageIncorrectBQChunkSize";
   public static final String ERROR_MSG_BQ_INCORRECT_TEMPORARY_BUCKET = "errorMessageIncorrectBQBucketName";
   public static final String ERROR_MSG_BQ_INCORRECT_PROPERTY = "errorMessageIncorrectBQProperty";
+  public static final String ERROR_MSG_INCORRECT_PARTITIONSTARTDATE = "errorMessageIncorrectPartitionStartDate";
```
Please check this change is not required.

itsmekumari commented on src/e2e-test/resources/pluginParameters.properties:
```diff
@@ -1,5 +1,6 @@
 projectId=cdf-athena
-datasetprojectId=cdf-athena
+datasetprojectId=testbq_bqmt
```
Why are we changing this? Please revert

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/utils/CdfPluginPropertyLocator.java:
```diff
@@ -74,6 +78,10 @@ public enum CdfPluginPropertyLocator {
       .put("createFailIfObjectExists", CdfPluginPropertyLocator.GCS_CREATE_FAIL_IF_OBJECT_EXISTS)
       .put("gcsMoveSourcePath", CdfPluginPropertyLocator.GCS_MOVE_SOURCE_PATH)
       .put("gcsMoveDestinationPath", CdfPluginPropertyLocator.GCS_MOVE_DESTINATION_PATH)
+      .put("filter", CdfPluginPropertyLocator.FILTER)
```
Use the latest existing steps mentioned earlier. This change is not required.

itsmekumari replied:
Recheck this

itsmekumari commented on src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:
```diff
@@ -90,18 +90,18 @@ public void getCountOfNoOfRecordsTransferredToTargetBigQueryTable() throws IOExc
     int countRecordsTarget = BigQueryClient.countBqQuery(TestSetupHooks.bqTargetTable);
     String selectQuery = "SELECT count(*)  FROM `" + projectId + "." + datasetName + "." +
-      TestSetupHooks.bqTargetTable + "` WHERE " + PluginPropertyUtils.pluginProp(filter);
+            TestSetupHooks.bqTargetTable + "` WHERE " + PluginPropertyUtils.pluginProp(filter);
```
Revert the unrequired changes, still I can see in the PR.

itsmekumari reviewed (changes requested):
Address all the review comments. And check the build is in green.

itsmekumari commented on src/e2e-test/resources/errorMessage.properties:
```diff
@@ -33,4 +33,7 @@ errorMessageMultipleFileWithoutClearDefaultSchema=Found a row with 4 fields when
+errorMessageIncorrectPartitionStartDate=02-01-2025 is not in a valid format. Enter valid date in format: yyyy-MM-dd
+errorMessageIncorrectPartitionEndDate=03-01-2025 is not in a valid format. Enter valid date in format: yyyy-MM-dd
+errorMessageIncorrectReferenceName=Invalid reference name 'invalidRef&^*&&*'. Supported characters are: letters, numbers, and '_', '-', '.', or '$'.
+errorLogsMessageInvalidFilter=Spark Program 'phase-1' failed.
```
As mentioned earlier also, add a relevant error message.
